ansible/roles/test/files/ptftests/ip_in_ip_tunnel_test.py

```diff
@@ -0,0 +1,162 @@
+
+'''
+#---------------------------------------------------------------------
+# Global imports
+#---------------------------------------------------------------------
+import logging
+import random
+from ipaddress import ip_address
+import ptf
+import ptf.packet as scapy
+from ptf.base_tests import BaseTest
+from ptf.mask import Mask
+from ptf.testutils import *
+
+PACKET_NUM = 10000
+DIFF = 0.25 # The valid range for balance check
+SRC_IP_RANGE = [unicode('8.0.0.0'), unicode('8.255.255.255')]
+TIMEOUT = 1
+
+class IpinIPTunnelTest(BaseTest):
+    '''
+    @summary: Overview of functionality
+        This script send traffic to standby ToR, and capture traffic
+         on all portchannel interfaces to check balance.
+    '''
+    def __init__(self):
+        '''
+        @summary: constructor
+        '''
+        BaseTest.__init__(self)
+        self.test_params = test_params_get()
+
+    def setUp(self):
+        self.server_ip = self.test_params['server_ip']
+        self.active_tor_mac = self.test_params['active_tor_mac']
+        self.standby_tor_mac = self.test_params['standby_tor_mac']
+        self.active_tor_ip = self.test_params['active_tor_ip']
+        self.standby_tor_ip = self.test_params['standby_tor_ip']
+        self.ptf_portchannel_indices = self.test_params['ptf_portchannel_indices']
+        self.indice_to_portchannel = {}
+        for port_channel, indices in self.ptf_portchannel_indices.items():
+            for indice in indices:
+                self.indice_to_portchannel[indice] = port_channel
+
+        self.hash_key_list = self.test_params['hash_key_list']
+        self.dataplane = ptf.dataplane_instance
+
+    def runTest(self):
+        """
+        Entrypoint of test script.
+        """
+        self.send_and_verify_packets()
+
+    def random_ip(self, begin, end):
+        """
+        Generate a random IP from given ip range
+        """
+        length = int(ip_address(end)) - int(ip_address(begin))
+        return str(ip_address(begin) + random.randint(0, length))
+
+    def generate_packet_to_server(self, hash_key):
+        """
+        Generate a packet to server. The value of field in packet is filled with random value according to hash_key
+        """
+        base_src_mac = self.dataplane.get_mac(0, 0)
+        ip_src = self.random_ip(SRC_IP_RANGE[0], SRC_IP_RANGE[1]) if hash_key == 'src-ip' else SRC_IP_RANGE[0]
+        ip_dst = self.server_ip
+        sport = random.randint(1, 65535) if hash_key == 'src-port' else 1234
```

Comment:
I'm not sure if this random generation will always provide consistent results. May be a rerun might pass. Could you check if the logic is similar to existing fib hash test?

Reply:
Yes. The logic is almost the same with fib hash test. Theoretically, if we generate enough packets, the distribution will be even enough.